File: rotate-image/Solution48.java
// https://leetcode.com/problems/rotate-image/

class Solution48 {

    // T: O(n^2), S: O(1)
    public void rotate(int[][] matrix) {
        int n = matrix.length;

        // Transpose (swap rows and cols) the matrix
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }

        // System.out.println("Transposed:");
        // printMatrix(matrix);

        // Reverse each row
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n / 2; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[i][n - 1 - j];
                matrix[i][n - 1 - j] = temp;
            }
        }
    }

    // T: O(n^2), S: O(n)
    public void rotateBrute(int[][] matrix) {
        int n = matrix.length;
        int[][] orig = new int[n][];
        for (int i = 0; i < n; i++) {
            orig[i] = matrix[i].clone();
        }
        for (int r = 0; r < n; r++) {
            int[] nums = orig[r];
            copyRowToCol(matrix, nums, n - r - 1);
        }
    }

    private void copyRowToCol(int[][] matrix, int[] nums, int col) {
        int n = matrix.length;
        for (int i = 0; i < n; i++) {
            matrix[i][col] = nums[i];
        }
    }

    // Helper method to print the matrix
    private void printMatrix(int[][] matrix) {
        for (int[] row : matrix) {
            for (int num : row) {
                System.out.print(num + " ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        var sol = new Solution48();

        int[][] matrix1 = {
                { 1, 2, 3 },
                { 4, 5, 6 },
                { 7, 8, 9 }
        };
        System.out.println("Original:");
        sol.printMatrix(matrix1);
        sol.rotate(matrix1);
        System.out.println("Rotated:");
        sol.printMatrix(matrix1);

        int[][] matrix2 = {
                { 5, 1, 9, 11 },
                { 2, 4, 8, 10 },
                { 13, 3, 6, 7 },
                { 15, 14, 12, 16 }
        };
        System.out.println("Original:");
        sol.printMatrix(matrix2);
        sol.rotate(matrix2);
        System.out.println("Rotated:");
        sol.printMatrix(matrix2);

        int[][] matrix3 = {
                { 1, 2 },
                { 3, 4 },
        };
        System.out.println("Original:");
        sol.printMatrix(matrix3);
        sol.rotate(matrix3);
        System.out.println("Rotated:");
        sol.printMatrix(matrix3);
    }

}

File: rotate-image/solution48.py
# https://leetcode.com/problems/rotate-image/

from typing import List


class Solution48:

    # T: O(n), S: O(1)
    def rotate(self, matrix: List[List[int]]) -> None:
        n = len(matrix)

        # Transpose the matrix
        for i in range(n):
            for j in range(i, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]

        # print("Transposed:")
        # self.print_matrix(matrix)

        # Reverse each row
        for i in range(n):
            for j in range(n // 2):
                matrix[i][j], matrix[i][n - 1 -
                                        j] = matrix[i][n - 1 - j], matrix[i][j]

    # T: O(n^2), S: O(1)
    def rotateBrute(self, matrix: List[List[int]]) -> None:
        n = len(matrix)
        orig = [row[:] for row in matrix]
        for r in range(n):
            nums = orig[r]
            self.copy_row_to_col(matrix, nums, n - r - 1)

    def copy_row_to_col(self, matrix: List[List[int]], nums: List[int], col):
        n = len(matrix)
        for i in range(n):
            matrix[i][col] = nums[i]

    def print_matrix(self, matrix: List[List[int]]) -> None:
        for row in matrix:
            print(" ".join(map(str, row)))


def main():
    sol = Solution48()

    matrix1 = [
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]
    ]
    print("Original:")
    sol.print_matrix(matrix1)
    sol.rotate(matrix1)
    print("Rotated:")
    sol.print_matrix(matrix1)

    matrix2 = [
        [5, 1, 9, 11],
        [2, 4, 8, 10],
        [13, 3, 6, 7],
        [15, 14, 12, 16]
    ]
    print("Original:")
    sol.print_matrix(matrix2)
    sol.rotate(matrix2)
    print("Rotated:")
    sol.print_matrix(matrix2)

    matrix3 = [
        [1, 2],
        [3, 4],
    ]
    print("Original:")
    sol.print_matrix(matrix3)
    sol.rotate(matrix3)
    print("Rotated:")
    sol.print_matrix(matrix3)


if __name__ == "__main__":
    main()
